File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/service/UserService.java
package com.windea.demo.cloudcollect.core.service;

import com.windea.demo.cloudcollect.core.domain.entity.*;
import com.windea.demo.cloudcollect.core.domain.enums.Role;
import com.windea.demo.cloudcollect.core.domain.view.EmailRegisterView;
import com.windea.demo.cloudcollect.core.domain.view.UsernamePasswordLoginView;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * 用户的服务（登录、重置密码等功能委托给{@code UserDetailsService}）。
 */
public interface UserService {
	/**
	 * 通过用户名&密码登录用户。
	 */
	User loginByUsernameAndPassword(UsernamePasswordLoginView view);

	/**
	 * 通过邮箱注册用户（密码需要加密）。
	 */
	void registerByEmail(EmailRegisterView view);

	/**
	 * 激活用户（将activated设为true）。
	 */
	void activate(User user);

	/**
	 * 重置用户密码（密码需要加密）。
	 */
	void resetPassword(User user, String newPassword);

	/**
	 * 更新用户信息（不允许同时修改密码）。
	 */
	void update(Long id, User user);

	/**
	 * 得到用户信息。
	 */
	User get(Long id);

	/**
	 * 分页得到某一用户的所有关注用户。
	 */
	Page<User> getFollowToUserPage(Long id, Pageable pageable);

	/**
	 * 得到某一用户的关注用户数量。
	 */
	Long getFollowToUserCount(Long id);

	/**
	 * 分页得到某一用户的所有粉丝用户。
	 */
	Page<User> getFollowByUserPage(Long id, Pageable pageable);

	/**
	 * 得到某一用户的粉丝用户数量。
	 */
	Long getFollowByUserCount(Long id);

	/**
	 * 分页得到某一用户的所有收藏。
	 */
	Page<Collect> getCollectPage(Long id, Pageable pageable);

	/**
	 * 得到某一用户的收藏数量。
	 */
	Long getCollectCount(Long id);

	/**
	 * 得到某一用户的所有收藏分类。
	 */
	Page<CollectCategory> getCollectCategoryPage(Long id, Pageable pageable);

	/**
	 * 得到某一用户的所有收藏分类数量。
	 */
	Long getCollectCategoryCount(Long id);

	/**
	 * 分页得到某一用户的所有通知。
	 */
	Page<Notice> getNoticePage(Long id, Pageable pageable);

	/**
	 * 得到某一用户的通知数量。
	 */
	Long getNoticeCount(Long id);

	/**
	 * 分页得到所有用户。
	 */
	Page<User> findAll(Pageable pageable);

	/**
	 * 根据昵称分页全局模糊查询用户。
	 */
	Page<User> findByNickname(String nickname, Pageable pageable);

	/**
	 * 根据身份分页全局查询用户。
	 */
	Page<User> findByRole(Role role, Pageable pageable);

	/**
	 * 检查某一用户是否已存在。
	 */
	boolean exists(User user);
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/service/UserService.kt
package com.windea.demo.cloudcollect.core.service

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.domain.request.*
import com.windea.demo.cloudcollect.core.domain.response.*
import com.windea.demo.cloudcollect.core.enums.*
import org.springframework.data.domain.*
import org.springframework.web.multipart.*

interface UserService {
	/**登录用户。*/
	fun login(form: LoginForm): UserDetailsVo
	
	/**注册用户。*/
	fun register(user: User)
	
	///**注销用户。*/
	//fun logout()
	
	/**激活用户。*/
	fun activate(username: String, activateCode: String)
	
	/**忘记用户密码。*/
	fun forgotPassword(email: String)
	
	/**重置用户密码。*/
	fun resetPassword(form: ResetPasswordForm, resetPasswordCode: String)
	
	/**更新用户信息。*/
	fun modify(id: Long, user: User)
	
	/**上传用户头像。*/
	fun uploadAvatar(id: Long, multipartFile: MultipartFile): String
	
	/**关注某一用户。*/
	fun follow(id: Long)
	
	/**取消关注某一用户。*/
	fun unfollow(id: Long)
	
	/**根据id得到用户。*/
	fun findById(id: Long): User
	
	/**根据用户名得到用户。*/
	fun findByUsername(username: String): User
	
	/**根据邮箱得到用户。*/
	fun findByEmail(email: String): User
	
	/**得到随机用户。*/
	fun findByRandom(): User
	
	/**得到所有用户。*/
	fun findAll(pageable: Pageable): Page<User>
	
	/**根据昵称全局模糊查询用户。*/
	fun findAllByNicknameContains(nickname: String, pageable: Pageable): Page<User>
	
	/**根据用户名全局模糊查询用户。*/
	fun findAllByUsernameContains(username: String, pageable: Pageable): Page<User>
	
	/**根据邮箱全局模糊查询用户。*/
	fun findAllByEmailContains(email: String, pageable: Pageable): Page<User>
	
	/**根据身份全局查询用户。*/
	fun findAllByRole(role: Role, pageable: Pageable): Page<User>
	
	/**根据关注用户id查询用户。*/
	fun findAllByFollowToUserId(followToUserId: Long, pageable: Pageable): Page<User>
	
	/**根据昵称和关注用户id模糊查询用户。*/
	fun findAllByNicknameContainsAndFollowToUserId(nickname: String, followToUserId: Long, pageable: Pageable): Page<User>
	
	/**根据粉丝用户id查询用户。*/
	fun findAllByFollowByUserId(followByUserId: Long, pageable: Pageable): Page<User>
	
	/**根据昵称和粉丝用户id模糊查询用户。*/
	fun findAllByNicknameContainsAndFollowByUserId(nickname: String, followByUserId: Long, pageable: Pageable): Page<User>
	
	/**根据点赞收藏id查询用户。*/
	fun findAllByPraiseToCollectId(praiseToCollectId: Long, pageable: Pageable): Page<User>
	
	/**检查某一用户是否已存在。*/
	fun existsByUsernameOrEmail(username: String, email: String): Boolean
}
